packages/itwin/tree-widget/src/components/trees/stateless/models-tree/HierarchyVisibilityHandler.ts

```diff
@@ -0,0 +1,877 @@
+      // doesn't mean all its subcategories are displayed - this call ensures that
+      viewport.changeCategoryDisplay(categoryId, true, true);
+    }
+  }
+
+  private changeCategoryState(props: ChangeCategoryStateProps): Observable<void> {
+    const result = defer(() => {
+      const viewport = this._props.viewport;
+      const { modelId, categoryId, on } = props;
+      return concat(
+        props.on && !viewport.view.viewsModel(modelId) ? this.showModelWithoutAnyCategoriesOrElements(modelId) : EMPTY,
+        defer(() => {
+          this.changeCategoryStateInViewportAccordingToModelVisibility(modelId, categoryId, on);
+          return this.clearAlwaysAndNeverDrawnElements(props);
+        }),
+      );
+    });
+
+    const ovr = this._props.overrides?.changeCategoryState;
+    return ovr ? from(ovr(this.createVoidOverrideProps(props, result))) : result;
+  }
+
+  /**
+   * Updates visibility of all grouping node's elements.
+   * @see `changeElementState`
+   */
+  private changeElementGroupingNodeState(node: GroupingHierarchyNode, on: boolean): Observable<void> {
+    const result = this.getGroupedElementIds(node).pipe(
+      mergeMap((group) => {
+        if (!group) {
+          return EMPTY;
+        }
+        const { modelId, categoryId, elementIds } = group;
+        const viewport = this._props.viewport;
+        return concat(
+          on && !viewport.view.viewsModel(modelId) ? from(viewport.addViewedModels(modelId)) : EMPTY,
+          defer(() => {
+            const categoryVisibility = this.getDefaultCategoryVisibilityStatus(categoryId, modelId);
+            const isDisplayedByDefault = categoryVisibility.state === "visible";
+            return from(elementIds).pipe(this.changeElementStateNoChildrenOperator({ on, isDisplayedByDefault }));
+          }),
+        );
+      }),
+    );
+
+    const ovr = this._props.overrides?.changeElementGroupingNodeState;
+    return ovr ? from(ovr(this.createVoidOverrideProps({ node, on }, result))) : result;
+  }
+
+  /**
+   * Updates visibility of an element and all its child elements by adding them to the always/never drawn list.
+   * @note If element is to be enabled and model is hidden, it will be enabled.
+   */
+  private changeElementState(props: ChangeElementStateProps): Observable<void> {
+    const result = defer(() => {
+      const { elementId, on, hasChildren, modelId, categoryId } = props;
+      const viewport = this._props.viewport;
+      return concat(
+        props.on && !viewport.view.viewsModel(modelId) ? this.showModelWithoutAnyCategoriesOrElements(modelId) : EMPTY,
+        defer(() => {
+          const categoryVisibility = this.getDefaultCategoryVisibilityStatus(categoryId, modelId);
+          const isDisplayedByDefault = categoryVisibility.state === "visible";
+          return of(elementId).pipe(
+            concatWith(hasChildren ? from(this._queryHandler.queryElements({ rootElementIds: props.elementId })) : EMPTY),
```

Comment:
I thought we talked about not doing this for child elements

Reply:
This is the original behavior for `ModelsVisibilityHandler`:
```ts
protected async changeElementState(id: Id64String, modelId: Id64String | undefined, categoryId: Id64String | undefined, on: boolean, hasChildren?: boolean) {
    const isDisplayedByDefault = this.isElementDisplayedByDefault(modelId, categoryId);
    const isHiddenDueToExclusiveAlwaysDrawnElements = this.hasExclusiveAlwaysDrawnElements();
    this.changeElementStateInternal(id, on, isDisplayedByDefault, isHiddenDueToExclusiveAlwaysDrawnElements);
    if (!hasChildren) {
      return;
    }

    const childIdsContainer = this.getAssemblyElementIds(id);
    for await (const childId of childIdsContainer.getElementIds()) {
      this.changeElementStateInternal(childId, on, isDisplayedByDefault, isHiddenDueToExclusiveAlwaysDrawnElements);
    }
  }
```

Reply:
Well, the original `ModelsVisibilityHandler` was very inconsistent with its visibility handling. IMO, if parent node's visibility state doesn't depend on children states, then changing visibility of the parent shouldn't affect children visibility either.

Reply:
Removed handling of element children